File: common_book/src/qbmfxsydq/java/com/intelligent/reader/fragment/WebViewFragment.java
package com.intelligent.reader.fragment;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.Context;
import android.os.Build;
import android.os.Bundle;
import android.os.Handler;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.text.TextUtils;
import android.util.DisplayMetrics;
import android.view.InflateException;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.ding.basic.request.RequestService;
import com.dingyue.contract.util.SharedPreUtil;
import com.intelligent.reader.BuildConfig;
import com.intelligent.reader.R;
import com.intelligent.reader.app.BookApplication;
import com.intelligent.reader.util.PagerDesc;
import com.intelligent.reader.view.SelectSexDialog;

import net.lzbook.kit.appender_loghub.StartLogClickUtil;
import net.lzbook.kit.book.view.LoadingPage;
import net.lzbook.kit.request.UrlUtils;
import net.lzbook.kit.utils.AppLog;
import net.lzbook.kit.utils.AppUtils;
import net.lzbook.kit.utils.CustomWebClient;
import net.lzbook.kit.utils.JSInterfaceHelper;

import java.lang.ref.WeakReference;
import java.util.HashMap;
import java.util.Map;

public class WebViewFragment extends Fragment implements SelectSexDialog.onAniFinishedCallback {

    private static String TAG = WebViewFragment.class.getSimpleName();
    public String url = "";
    private String type = "";
    private WeakReference<Activity> weakReference;
    private Context context;
    private View rootView;
    private RelativeLayout contentLayout;
    private WebView contentView;
    private CustomWebClient customWebClient;
    private JSInterfaceHelper jsInterfaceHelper;
    private FragmentCallback fragmentCallback;
    private LoadingPage loadingpage;
    private Handler handler;
    private TextView txt_head_title;
    private int bottomType;//青果打点搜索 2 推荐  3 榜单
    private RelativeLayout rl_head;
    private View img_shadow;
    private int count = 0;
    private SelectSexDialog selectSexDialog;
    private SharedPreUtil sharedPreUtil;
    private ImageView img_sex;
    public PagerDesc mPagerDesc;
    private int h5Margin;

    @Override
    public void onAttach(Activity activity) {
        super.onAttach(activity);
        this.weakReference = new WeakReference<>(activity);
        this.fragmentCallback = (FragmentCallback) activity;
        this.context = activity;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        handler = new Handler();
        AppLog.e(TAG, "----------->start");
        Bundle bundle = this.getArguments();
        if (bundle != null) {
            this.url = bundle.getString("url");
            AppLog.e(TAG, "url---->" + url);
            type = bundle.getString("type");
        }
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
            Bundle savedInstanceState) {
        AppLog.e(TAG, "onCreateView");
        try {
            rootView = inflater.inflate(R.layout.webview_layout, container, false);
        } catch (InflateException e) {
            e.printStackTrace();
        }
        if (weakReference != null) {
            AppUtils.disableAccessibility(weakReference.get());
        }
        sharedPreUtil = new SharedPreUtil(SharedPreUtil.SHARE_DEFAULT);
        initView();
        return rootView;
    }

    public void setTitle(String title, int logBottomType) {
        if (txt_head_title != null) {
            txt_head_title.setText(title);
        }
        bottomType = logBottomType;
    }

    @SuppressLint("JavascriptInterface")
    private void initView() {
        if (rootView != null) {
            contentLayout = (RelativeLayout) rootView.findViewById(R.id.web_content_layout);
            contentView = (WebView) rootView.findViewById(R.id.web_content_view);
            txt_head_title = rootView.findViewById(R.id.txt_head_title);
            rl_head = (RelativeLayout) rootView.findViewById(R.id.rl_head);
            img_sex = rootView.findViewById(R.id.img_sex);
            img_shadow = rootView.findViewById(R.id.img_shadow);

            if (contentView != null) {
                contentView.setLayerType(View.LAYER_TYPE_SOFTWARE, null);
            }
            if(img_sex != null){
                img_sex.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {

                        if(weakReference.get() != null){
                            if (selectSexDialog == null) {
                                selectSexDialog = new SelectSexDialog(weakReference.get());
                                selectSexDialog.setAniFinishedAction(WebViewFragment.this);
                            }
                            Map<String,String> data = new HashMap<>();
                            //0 表示男  1 表示女
                            if (sharedPreUtil.getInt(SharedPreUtil.RANK_SELECT_SEX, 0) == 0) {
                                data.put("type","2");
                                sharedPreUtil.putInt(SharedPreUtil.RANK_SELECT_SEX, 1);
                                img_sex.setImageResource(R.drawable.rank_gril_icon);
                                selectSexDialog.show(false);
                                String uri = RequestService.WEB_RANK_H5_Girl.replace("{packageName}",
                                        AppUtils.getPackageName());
                                url = UrlUtils.buildWebUrl(uri, new HashMap());
                                loadingData(url);
                            } else {
                                data.put("type","1");
                                sharedPreUtil.putInt(SharedPreUtil.RANK_SELECT_SEX, 0);
                                selectSexDialog.show(true);
                                img_sex.setImageResource(R.drawable.rank_boy_icon);
                                String uri = RequestService.WEB_RANK_H5_BOY.replace("{packageName}",
                                        AppUtils.getPackageName());
                                url = UrlUtils.buildWebUrl(uri, new HashMap());
                                loadingData(url);
                            }
                            StartLogClickUtil.upLoadEventLog(weakReference.get(),StartLogClickUtil.TOP_PAGE,StartLogClickUtil.QG_SWITCHTAB,data);
                        }

                    }
                });
            }

        }

        if (type.equals("recommend_male") || type.equals("recommend_female")) {
            if (rl_head != null) {
                rl_head.setVisibility(View.GONE);
            }
            if (img_shadow != null) {
                img_shadow.setVisibility(View.GONE);
            }
        }
        if(img_sex != null){
            if ("rankBoy".equals(type)) {
                img_sex.setVisibility(View.VISIBLE);
                img_sex.setImageResource(R.drawable.rank_boy_icon);
            } else if ("rankGirl".equals(type)) {
                img_sex.setVisibility(View.VISIBLE);
                img_sex.setImageResource(R.drawable.rank_gril_icon);
            } else {
                img_sex.setVisibility(View.GONE);
            }
        }

        if (weakReference != null) {
            loadingpage = new LoadingPage(weakReference.get(), contentLayout);
        }

        if (contentView != null && context != null) {
            customWebClient = new CustomWebClient(context, contentView);
        }

        if (contentView != null && customWebClient != null) {
            customWebClient.setWebSettings();
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                contentView.getSettings().setMixedContentMode(
                        WebSettings.MIXED_CONTENT_ALWAYS_ALLOW);
            }
            contentView.setWebViewClient(customWebClient);
        }

        if (contentView != null && context != null) {
            jsInterfaceHelper = new JSInterfaceHelper(context, contentView);
        }

        if (jsInterfaceHelper != null && contentView != null) {
            contentView.addJavascriptInterface(jsInterfaceHelper, "J_search");
        }

        if (fragmentCallback != null && jsInterfaceHelper != null) {
            fragmentCallback.webJsCallback(jsInterfaceHelper);
        }
        addTouchListener();
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        if (!TextUtils.isEmpty(url)) {
            loadWebData(url);
        }
    }

    @Override
    public void onResume() {
        super.onResume();
    }

    @Override
    public void onPause() {
        super.onPause();
    }

    public void loadWebData(String url) {
        AppLog.e(TAG, "loadWebData url: " + url);
        if (fragmentCallback != null && contentView != null) {
            url = fragmentCallback.startLoad(contentView, url);
        }
        AppLog.e(TAG, "loadWebData url: " + url);
        startLoading(handler, url);
        webViewCallback();
    }

    private void startLoading(Handler handler, final String url) {
        if (contentView == null) {
            return;
        }

        if (handler != null) {
            handler.post(new Runnable() {
                @Override
                public void run() {
                    loadingData(url);
                }
            });
        } else {
            loadingData(url);
        }
    }

    public boolean isNeedInterceptSlide() {
        if (this.url.contains("recommend")) {
            return true;
        }
        return false;
    }

    private void addTouchListener() {
        if (contentView != null && isNeedInterceptSlide()) {
            contentView.setOnTouchListener(new View.OnTouchListener() {
                @Override
                public boolean onTouch(View v, MotionEvent event) {
                    float y = event.getRawY();
                    switch (event.getAction()) {
                        case MotionEvent.ACTION_DOWN:
                            if (contentView != null) {
                                int[] loction = new int[2];
                                contentView.getLocationOnScreen(loction);
                                h5Margin = loction[1];
                            }
                            if (null != mPagerDesc) {
                                float top = mPagerDesc.getTop();
                                float bottom = top + (mPagerDesc.getBottom() - mPagerDesc.getTop());
                                DisplayMetrics metric = getResources().getDisplayMetrics();
                                top = (float) (top * metric.density) + h5Margin;
                                bottom = (float) (bottom * metric.density) + h5Margin;
                                if (y > top && y < bottom) {
                                    contentView.requestDisallowInterceptTouchEvent(true);
                                } else {
                                    contentView.requestDisallowInterceptTouchEvent(false);
                                }
                            }
                            break;
                        case MotionEvent.ACTION_UP:
                            break;
                        case MotionEvent.ACTION_MOVE:
                            break;
                        default:
                            break;
                    }
                    return false;
                }
            });
        }
    }


    private void loadingData(String url) {
        if (customWebClient != null) {
            customWebClient.doClear();
        }
        if (!TextUtils.isEmpty(url) && contentView != null) {
            try {
                AppLog.e(TAG, "WebViewFragment LoadingData ==> " + url);
                contentView.loadUrl(url);
            } catch (NullPointerException e) {
                e.printStackTrace();
                weakReference.get().finish();
            }
        }
    }

    private void webViewCallback() {
        if (rootView == null) {
            return;
        }

        if (customWebClient != null) {
            customWebClient.setStartedAction(new CustomWebClient.onStartedCallback() {
                @Override
                public void onLoadStarted(String url) {
                    AppLog.e(TAG, "onLoadStarted: " + url);
                }
            });

            customWebClient.setErrorAction(new CustomWebClient.onErrorCallback() {
                @Override
                public void onErrorReceived() {
                    AppLog.e(TAG, "onErrorReceived");
                    if (loadingpage != null) {
                        loadingpage.onErrorVisable();
                    }
                }
            });

            customWebClient.setFinishedAction(new CustomWebClient.onFinishedCallback() {
                @Override
                public void onLoadFinished() {
                    AppLog.e(TAG, "onLoadFinished");
                    if (loadingpage != null) {
                        loadingpage.onSuccessGone();
                    }

                }
            });
        }

        if (loadingpage != null) {
            loadingpage.setReloadAction(new LoadingPage.reloadCallback() {
                @Override
                public void doReload() {
                    AppLog.e(TAG, "doReload");
                    if (customWebClient != null) {
                        customWebClient.doClear();
                    }
                    if(contentView != null){
                        contentView.loadUrl(url);
                    }

                }
            });
        }

    }


    @Override
    public void onDestroy() {
        super.onDestroy();
        if (selectSexDialog != null) {
            selectSexDialog = null;
        }
        if (contentView != null) {
            contentView.clearCache(true); //清空缓存
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                if (contentLayout != null) {
                    contentLayout.removeView(contentView);

                }
                contentView.stopLoading();
                contentView.removeAllViews();
                contentView.destroy();
            } else {
                contentView.stopLoading();
                contentView.removeAllViews();
                contentView.destroy();
                if (contentLayout != null) {
                    contentLayout.removeView(contentView);
                }
            }
            contentView = null;
        }
        if (BuildConfig.DEBUG) {
            BookApplication.getRefWatcher().watch(this);
        }

    }

    @Override
    public void onAniFinished() {
        if (weakReference.get() != null && selectSexDialog != null) {
            if (selectSexDialog.isShow()) {
                selectSexDialog.dismiss();
            }
        }
    }

    public interface FragmentCallback {
        void webJsCallback(JSInterfaceHelper jsInterfaceHelper);

        String startLoad(WebView webView, String url);

    }
}

File: common_book/src/qbmfkdxs/java/com/intelligent/reader/fragment/WebViewFragment.kt
package com.intelligent.reader.fragment

import android.annotation.SuppressLint
import android.content.Intent
import android.graphics.RectF
import android.os.Build
import android.os.Bundle
import android.os.Handler
import android.support.v4.app.Fragment
import android.text.TextUtils
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.webkit.JavascriptInterface
import android.webkit.WebSettings
import com.ding.basic.net.Config
import com.dingyue.searchbook.activity.SearchBookActivity
import com.google.gson.Gson
import com.intelligent.reader.R
import com.orhanobut.logger.Logger
import kotlinx.android.synthetic.qbmfkdxs.frag_web_view.*
import net.lzbook.kit.app.base.BaseBookApplication
import net.lzbook.kit.appender_loghub.StartLogClickUtil
import net.lzbook.kit.ui.widget.LoadingPage
import net.lzbook.kit.utils.AppUtils
import net.lzbook.kit.utils.NetWorkUtils
import net.lzbook.kit.utils.oneclick.OneClickUtil
import net.lzbook.kit.utils.router.RouterConfig
import net.lzbook.kit.utils.router.RouterUtil
import net.lzbook.kit.utils.runOnMain
import net.lzbook.kit.utils.web.CustomWebClient
import net.lzbook.kit.utils.web.JSInterfaceObject

open class WebViewFragment : Fragment(), View.OnClickListener {

    private var url: String? = ""
    private var type: String? = null

    private var customWebClient: CustomWebClient? = null
    private var jsInterfaceObject: JSInterfaceObject? = null

    private var loadingPage: LoadingPage? = null

    private var viewVisible: Boolean = false
    private var viewPrepared: Boolean = false

    val TYPE_RECOMM = "recommend"
    val TYPE_RANK = "rank"
    val TYPE_CATEGORY = "category"
    private var handle = Handler()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val bundle = this.arguments

        if (bundle != null) {
            this.url = bundle.getString("url")
            this.type = bundle.getString("type")
        }
    }

    private fun setTitle() {
        if (type?.contains(TYPE_RECOMM) == false) {
            rl_head.visibility = View.VISIBLE
            if (TYPE_RANK == type) {
                txt_title?.text = "榜单"
            } else if (TYPE_CATEGORY == type) {
                txt_title?.text = "分类"
            }
        } else {
            rl_head.visibility = View.GONE
        }

    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.frag_web_view, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        AppUtils.disableAccessibility(requireActivity())
        initParameter()

        viewPrepared = true
        img_search?.setOnClickListener(this)

        if (!TextUtils.isEmpty(url)) {
            when (type) {
                "recommend" -> {
                    requestWebViewData(url)
                }
                "recommendMan" -> {
                    requestWebViewData(url)
                }
                "recommendWoman" -> {
                    requestWebViewData(url)
                }
                "recommendFinish" -> {
                    handle.postDelayed({
                        requestWebViewData(url)
                    }, 2000)
                }
                "recommendFantasy" -> {
                    handle.postDelayed({
                        requestWebViewData(url)
                    }, 2000)
                }
                "recommendModern" -> {
                    handle.postDelayed({
                        requestWebViewData(url)
                    }, 2000)
                }
                "rank" -> {
                    handle.postDelayed({
                        requestWebViewData(url)
                    }, 2000)
                }
                "category" -> {
                    handle.postDelayed({
                        requestWebViewData(url)
                    }, 2000)
                }
            }
        }
        setTitle()
    }


    override fun setUserVisibleHint(isVisibleToUser: Boolean) {
        super.setUserVisibleHint(isVisibleToUser)
        if (userVisibleHint) {
            viewVisible = true
            contentViewVisible()
        } else {
            viewVisible = false
        }
    }

    override fun onClick(v: View) {
        when (v.id) {
            R.id.img_search -> {
                try {
                    var pageCode = StartLogClickUtil.RECOMMEND_PAGE
                    if (!TextUtils.isEmpty(url)) {
                        when (type) {
                            "recommend", "recommendMan", "recommendWoman", "recommendFinish", "recommendFantasy", "recommendModern" -> {
                                pageCode = StartLogClickUtil.RECOMMEND_PAGE
                            }
                            "rank" -> {
                                pageCode = StartLogClickUtil.TOP_PAGE
                            }
                            "category" -> {
                                pageCode = StartLogClickUtil.CLASS_PAGE
                            }
                        }
                    }

                    StartLogClickUtil.upLoadEventLog(requireContext(), pageCode, StartLogClickUtil.SEARCH)
                    val intent = Intent(activity, SearchBookActivity::class.java)
                    startActivity(intent)
                } catch (exception: Exception) {
                    exception.printStackTrace()
                }
            }
        }
    }

    @SuppressLint("AddJavascriptInterface", "JavascriptInterface")
    private fun initParameter() {

        web_view_content?.setLayerType(View.LAYER_TYPE_NONE, null)

        loadingPage = LoadingPage(requireActivity(), rl_web_view_root)

        customWebClient = CustomWebClient(requireContext(), web_view_content)

        customWebClient?.initWebViewSetting()

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            web_view_content?.settings?.mixedContentMode = WebSettings.MIXED_CONTENT_ALWAYS_ALLOW
        }

        web_view_content?.webViewClient = customWebClient


        jsInterfaceObject = object : JSInterfaceObject(requireActivity()) {


            @JavascriptInterface
            override fun startSearchActivity(data: String?) {

            }

            @JavascriptInterface
            override fun startTabulationActivity(data: String?) {
                if (data != null && data.isNotEmpty() && !activity.isFinishing) {
                    if (OneClickUtil.isDoubleClick(System.currentTimeMillis())) {
                        return
                    }

                    try {
                        val redirect = Gson().fromJson(data, JSRedirect::class.java)

                        if (redirect?.url != null && redirect.title != null) {
                            val bundle = Bundle()
                            bundle.putString("url", Config.webViewBaseHost + redirect.url)
                            bundle.putString("title", redirect.title)

                            if (redirect.from != null && (redirect.from?.isNotEmpty() == true)) {
                                when {
                                    redirect.from == "recommend" -> bundle.putString("from", "recommend")
                                    redirect.from == "ranking" -> bundle.putString("from", "ranking")
                                    redirect.from == "category" -> bundle.putString("from", "category")
                                    else -> bundle.putString("from", "other")
                                }
                            } else {
                                bundle.putString("from", "other")
                            }

                            RouterUtil.navigation(activity, RouterConfig.TABULATION_ACTIVITY, bundle)
                        }
                    } catch (exception: Exception) {
                        exception.printStackTrace()
                    }
                }
            }

            @JavascriptInterface
            override fun handleBackAction() {

            }

            @JavascriptInterface
            override fun hideWebViewLoading() {
                runOnMain {
                    loadingPage?.onSuccessGone()
                }
            }

            @JavascriptInterface
            override fun handleWebRequestResult(method: String?) {
                if (null != web_view_content) {
                    Logger.e("WebViewMethod: $method")
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
                        web_view_content.evaluateJavascript(method) { value -> Logger.e("ReceivedValue: $value") }
                    } else {
                        web_view_content.loadUrl(method)
                    }
                }
            }
        }

        web_view_content?.addJavascriptInterface(jsInterfaceObject, "J_search")

        web_view_content.addJavascriptInterface(JsPositionInterface(), "J_position")
    }

    private fun contentViewVisible() {
        if (type != null) {
            if (type == "rank") {
                notifyWebLog()
            }
            if (type == "recommend") {
                notifyWebLog()
            }
            if (type == "category") {
                notifyWebLog()
            }
        }
    }

    private fun notifyWebLog() {
        if (!viewVisible || !viewPrepared) {
            return
        }
    }

    private fun requestWebViewData(url: String?) {
        startLoadingWebViewData(url)
        initWebViewCallback()
    }

    private fun startLoadingWebViewData(url: String?) {
        if (web_view_content == null) {
            return
        }
        web_view_content?.post {
            handleLoadingWebViewData(url)
        }
    }

    private fun handleLoadingWebViewData(url: String?) {
        customWebClient?.initParameter()

        if (url != null && url.isNotEmpty()) {
            try {
                web_view_content?.loadUrl(url)
            } catch (exception: NullPointerException) {
                exception.printStackTrace()
                requireActivity().finish()
            }
        }
    }

    private fun initWebViewCallback() {
        if (web_view_content == null) {
            return
        }

        customWebClient?.setLoadingWebViewStart {
            Logger.e("LoadingWebView: $it")
        }

        customWebClient?.setLoadingWebViewError {
            loadingPage?.onErrorVisable()
        }

        customWebClient?.setLoadingWebViewFinish {
            //无网无缓存（error）
            val isOfflineNotStorage = jsInterfaceObject?.isOfflineNotStorage() ?: false
            if (!NetWorkUtils.isNetworkAvailable(BaseBookApplication.getGlobalContext()) && isOfflineNotStorage) {
                loadingPage?.onErrorVisable()
            } else {
                loadingPage?.onSuccessGone()
            }
        }

        loadingPage?.setReloadAction(LoadingPage.reloadCallback {
            customWebClient?.initParameter()
            web_view_content?.reload()
        })
    }

    override fun onDestroy() {
        super.onDestroy()
        if (web_view_content != null) {
            web_view_content?.clearCache(true)
            if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.LOLLIPOP) {
                rl_web_view_root?.removeView(web_view_content)
                web_view_content?.stopLoading()
                web_view_content?.settings?.javaScriptEnabled = false
                web_view_content?.clearHistory()
                web_view_content?.removeAllViews()
                web_view_content?.destroy()
            } else {
                web_view_content?.stopLoading()
                web_view_content?.settings?.javaScriptEnabled = false
                web_view_content?.clearHistory()
                web_view_content?.removeAllViews()
                web_view_content?.destroy()
                rl_web_view_root?.removeView(web_view_content)
            }
        }
    }

    /**
     * 获取web中banner的位置js回调
     */
    inner class JsPositionInterface {

        @JavascriptInterface
        fun insertProhibitSlideArea(x: String, y: String, width: String, height: String) {
            try {
                val viewWidth = java.lang.Float.parseFloat(width)
                val viewHeight = java.lang.Float.parseFloat(height)
                val scale = web_view_content.resources.displayMetrics.widthPixels / viewWidth

                web_view_content.insertProhibitSlideArea(RectF(
                        x.toFloat() * scale, y.toFloat() * scale, (x.toFloat() + viewWidth) * scale,
                        (y.toFloat() + viewHeight) * scale))
            } catch (exception: Exception) {
                exception.printStackTrace()
            }
        }
    }

}